File: src/rate-limiter/scripts/sliding-window.lua
-- Sliding window rate limiter (atomic)
--
-- How it works:
--   We store each request as a member in a sorted set, with its
--   timestamp (in milliseconds) as the score. To check the limit:
--   1. Remove all entries older than the window
--   2. Count remaining entries
--   3. If under the limit, add the new request
--   4. Refresh the key's TTL so it self-cleans
--
-- KEYS[1] = the sorted set key (e.g. "rl:user:42")
-- ARGV[1] = window start timestamp in ms (now - windowMs)
-- ARGV[2] = current timestamp in ms (now) — used as the score
-- ARGV[3] = limit (max requests per window)
-- ARGV[4] = unique request ID (score:random, to avoid collisions)
-- ARGV[5] = key TTL in seconds (= window duration)
--
-- Returns: { allowed (0/1), count after operation }

-- Step 1: Remove entries that have fallen outside the window.
-- '-inf' to windowStart covers everything older than our window.
redis.call('ZREMRANGEBYSCORE', KEYS[1], '-inf', ARGV[1])

-- Step 2: How many requests are currently in the window?
local count = redis.call('ZCARD', KEYS[1])

local allowed = 0

-- Step 3: If under the limit, add this request to the set.
-- The score is the current timestamp; the member is a unique ID
-- so that two requests at the same millisecond don't collide.
if count < tonumber(ARGV[3]) then
    redis.call('ZADD', KEYS[1], ARGV[2], ARGV[4])
    allowed = 1
    count = count + 1
end

-- Step 4: Set a TTL on the key so it eventually cleans itself up
-- even if no more requests come in. We set it to the full window
-- duration — any entry in the set will expire within this time.
redis.call('EXPIRE', KEYS[1], ARGV[5])

return { allowed, count }

File: src/rate-limiter/strategies/sliding-window.ts
import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import type { Redis } from "ioredis";
import type { RateLimitResult } from "../index.js";
import type { RateLimitStrategy } from "./types.js";

// Load the Lua script once at module load time.
// In production, ioredis will cache it via EVALSHA after the first call.
const __dirname = dirname(fileURLToPath(import.meta.url));
const SCRIPT = readFileSync(
	join(__dirname, "..", "scripts", "sliding-window.lua"),
	"utf-8",
);

/**
 * Generate a unique member ID for each request.
 * We use timestamp:random to avoid sorted set member collisions —
 * if two requests arrive at the exact same millisecond, the random
 * suffix ensures they're stored as separate entries.
 */
function requestId(nowMs: number): string {
	return `${nowMs}:${Math.random().toString(36).slice(2, 10)}`;
}

export function createSlidingWindow(
	redis: Redis,
	limit: number,
	windowSec: number,
): RateLimitStrategy {
	const windowMs = windowSec * 1000;

	return {
		async limit(key: string): Promise<RateLimitResult> {
			const nowMs = Date.now();
			const windowStart = nowMs - windowMs;

			// eval(script, numKeys, ...keys, ...args)
			// Returns [allowed (0/1), count]
			const [allowed, count] = (await redis.eval(
				SCRIPT,
				1,
				key,
				windowStart.toString(),
				nowMs.toString(),
				limit.toString(),
				requestId(nowMs),
				windowSec.toString(),
			)) as [number, number];

			return {
				allowed: allowed === 1,
				remaining: Math.max(0, limit - count),
				limit,
				retryAfter: allowed === 1 ? 0 : windowSec,
			};
		},

		async peek(key: string): Promise<RateLimitResult> {
			const nowMs = Date.now();
			const windowStart = nowMs - windowMs;

			// Clean up expired entries, then count — but don't add anything.
			// We use a pipeline (two commands) rather than Lua here
			// because there's no conditional logic needed for a read-only peek.
			const pipeline = redis.pipeline();
			pipeline.zremrangebyscore(key, "-inf", windowStart);
			pipeline.zcard(key);
			const results = await pipeline.exec();

			// pipeline.exec() returns [[err, result], [err, result], ...]
			const count = (results![1][1] as number) ?? 0;

			const allowed = count < limit;
			return {
				allowed,
				remaining: Math.max(0, limit - count),
				limit,
				retryAfter: allowed ? 0 : windowSec,
			};
		},

		async reset(key: string): Promise<void> {
			await redis.del(key);
		},
	};
}
